core/drivers/ls_i2c.c

```diff
@@ -0,0 +1,423 @@
+			/* Write 1 to clear the IBIF field */
+			io_write8((vaddr_t)&regs->ibsr, reg);
+			break;
+		}
+		mdelay(1);
+	}
+
+	if (index == I2C_NUM_RETRIES)
+		return TEE_ERROR_BUSY;
+
+	if (test_rx_ack && (reg & I2C_IBSR_RXAK))
+		return TEE_ERROR_NO_DATA;
+
+	if (reg & I2C_IBSR_TCF)
+		return TEE_SUCCESS;
+
+	return TEE_ERROR_GENERIC;
+}
+
+static TEE_Result i2c_read(struct i2c_regs *regs, uint32_t slave_address,
+			   struct i2c_operation *operation,
+			   bool is_last_operation)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t index = 0;
+
+	/* Write Slave Address */
+	io_write8((vaddr_t)&regs->ibdr, (slave_address << BIT(0)) | BIT(0));
+	res = i2c_transfer_complete(regs, I2C_BUS_TEST_RX_ACK);
+	if (res)
+		return res;
+
+	/* select Receive mode. */
+	i2c_io_and8((vaddr_t)&regs->ibcr, ~I2C_IBCR_TXRX);
+	if (operation->length_in_bytes > 1) {
+		/* Set No ACK = 0 */
+		i2c_io_and8((vaddr_t)&regs->ibcr, ~I2C_IBCR_NOACK);
+	}
+
+	/* Perform a dummy read to initiate the receive operation. */
+	io_read8((vaddr_t)&regs->ibdr);
+
+	for (index = 0; index < operation->length_in_bytes; index++) {
+		res = i2c_transfer_complete(regs, I2C_BUS_NO_TEST_RX_ACK);
+		if (res)
+			return res;
+		if (index == (operation->length_in_bytes - 2)) {
+			/* Set No ACK = 1 */
+			i2c_io_or8((vaddr_t)&regs->ibcr, I2C_IBCR_NOACK);
+		} else if (index == (operation->length_in_bytes - 1)) {
+			if (!is_last_operation) {
+				/* select Transmit mode (for repeat start) */
+				i2c_io_or8((vaddr_t)&regs->ibcr, I2C_IBCR_TXRX);
+			} else {
+				/* Generate Stop Signal */
+				i2c_io_and8((vaddr_t)&regs->ibcr,
+					    ~(I2C_IBCR_MSSL | I2C_IBCR_TXRX));
+				res = i2c_bus_test_bus_busy
+					(regs, I2C_BUS_TEST_IDLE);
+				if (res)
+					return res;
+			}
+		}
+		operation->buffer[index] = io_read8((vaddr_t)&regs->ibdr);
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_write(struct i2c_regs *regs, uint32_t slave_address,
+			    struct i2c_operation *operation)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t index = 0;
+
+	/* Write Slave Address */
+	io_write8((vaddr_t)&regs->ibdr,
+		  (slave_address << BIT(0)) & (uint8_t)(~(BIT(0))));
+	res = i2c_transfer_complete(regs, I2C_BUS_TEST_RX_ACK);
+	if (res)
+		return res;
+
+	/* Write Data */
+	for (index = 0; index < operation->length_in_bytes; index++) {
+		io_write8((vaddr_t)&regs->ibdr, operation->buffer[index]);
+		res = i2c_transfer_complete(regs, I2C_BUS_TEST_RX_ACK);
+		if (res)
+			return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_stop(struct i2c_regs *regs)
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint8_t reg = 0;
+
+	reg = io_read8((vaddr_t)&regs->ibsr);
+	if (reg & I2C_IBSR_IBB) {
+		/* Generate Stop Signal */
+		i2c_io_and8((vaddr_t)&regs->ibcr,
+			    ~(I2C_IBCR_MSSL | I2C_IBCR_TXRX));
+		res = i2c_bus_test_bus_busy(regs, I2C_BUS_TEST_IDLE);
+		if (res)
+			return res;
+	}
+
+	/* Disable I2c Controller */
+	i2c_io_or8((vaddr_t)&regs->ibcr, I2C_IBCR_MDIS);
+
+	return res;
```

Comment:
We can only reach this point if `res` is `TEE_SUCCESS`. By explicitly returning `TEE_SUCCESS` it would be more clear that we're good if we've reached this point. Besides it's a different pattern compared to the functions above.